File: 16.Hashmap/basic.java
import java.util.HashMap;
import java.util.ArrayList;

public class basic{

	public static void main(String[] args)
	{
		hashmap();
		String s="abbcdabbbbabbbacde";
		frequency_map(s);
		character_index(s);
	}

	public static void character_index(String s)
	{
		Hashmap<Character,ArrayList<Integer>> map=new Hashmap<>();
		
	}

	public static void hashmap()
	{
		HashMap<String,Integer> map=new HashMap<>();
		map.put("INIDA",1000);
		map.put("delhi",100);
		map.put("russia",10);
		map.put("usa",1);

		//print
		System.out.println(map);

		//keyset
		ArrayList<String> keys=new ArrayList<>(map.keySet());
		System.out.println(keys);

		for(String s:map.keySet())
			System.out.println(s);

		//get
		System.out.println(map.get("Usa")); // if it is not present than it will give output null hence we use getOrDefault
		System.out.println(map.getOrDefault("Usa",0));

	}

	public static void frequency_map(String s)
	{	
		HashMap<Character,Integer> map=new HashMap<>();
		for(int i=0;i<s.length();i++)
		{
			Character ch=s.charAt(i);
			map.put(ch,map.getOrDefault(ch,0)+1);
			
		}

		System.out.println(map);
		return;
	}

}

File: 16.Hashmap/basic.cpp
#include<iostream>
#include<unordered_map>
#include<vector>
using namespace std;
void map1_basic()
{
	unordered_map<string,int> map;
	
	//put
	map["ind"]=10;
	map["usa"]=100;
	map["russia"]=1000;

	//get
	cout<<map["shubham"]<<endl;  // never do this as it will put shubham in hashmap too hence use map.count;
	cout<<map.count("shubham")<<endl; // it returns 0/1 if present or not

	//erase
	if(map.count("shubham")) map.erase("shubham");
	if(map.erase("delhi")) map.erase("delhi"); // first check whether it is present in hashmap or not then erase

	for(pair<string,int> p: map)
	{
		cout<<p.first<<" "<<p.second<<endl;
		map[p.first]+=100; // it will increase the value by 100;
	}
	for(pair<string,int> p: map)
	{
		cout<<p.first<<" "<<p.second<<endl;
		
	}
	return ;
}
void frequency_map(string&s)
{	
	unordered_map<char,int> map;
	for(int i=0;i<s.length();i++)
	{
		map[s[i]]++;
	}
	for(pair<char,int> p:map)
		cout<<p.first<<" "<<p.second<<endl;

	return;
}
void character_index(string s)
{
	unordered_map<char,vector<int>> map;
	for(int i=0;i<s.length();i++)
	{
		char ch=s[i];
		map[ch].push_back(i);
	}
	for(pair<char,vector<int>> p: map){
		cout<<p.first<<"->";
		  for(int i=0;i<p.second.size();i++) 
		  	cout<<p.second[i]<<", ";
		cout<<endl;
	}
}
int main()
{
	//map1_basic();
	string s="abbcdabbbbabbbacde";
	//frequency_map(s);
	character_index(s);
	return 0;
}